src/utils.test.js

```diff
@@ -106,22 +106,79 @@ describe('xeqBy', () => {
-  Object.defineProperty(document.documentElement, 'scrollWidth', { value: 200 })
-  Object.defineProperty(document.documentElement, 'clientWidth', { value: 100 })
+  Object.defineProperties(document.documentElement, {
+    scrollLeft: { value: 50 },
+    scrollWidth: { value: 200 },
+    clientWidth: { value: 100 }
+  })
 
   it('returns the horizontal scroll percentage', () => {
     expect(scrollPercentX()).toBe(0.5)
   })
 })
 
 describe('scrollPercentY', () => {
-  Object.defineProperty(document.documentElement, 'scrollTop', { value: 50 })
-  Object.defineProperty(document.documentElement, 'scrollHeight', { value: 200 })
-  Object.defineProperty(document.documentElement, 'clientHeight', { value: 100 })
+  describe('scrollTop is half way through the document', () => {
+    it('returns 0.5', () => {
+      Object.defineProperties(document.documentElement, {
+        scrollTop: { value: 50, configurable: true },
+        scrollHeight: { value: 200, configurable: true },
+        clientHeight: { value: 100, configurable: true }
+      })
+
+      expect(scrollPercentY()).toBe(0.5)
+    })
+  })
 
-  it('returns the vertical scroll percentage', () => {
-    expect(scrollPercentY()).toBe(0.5)
+  // Webkit/iOS can return scroll values less than zero
+  describe('scrollTop has a negative value', () => {
+    it('clamps the return value to 0', () => {
+      Object.defineProperties(document.documentElement, {
+        scrollTop: { value: -5, configurable: true },
+        scrollHeight: { value: 200, configurable: true },
+        clientHeight: { value: 100, configurable: true }
+      })
+
+      expect(scrollPercentY()).toBe(0)
+    })
+  })
+
+  // Webkit/iOS can return scroll values greater than the
+  // total height of the document
+  describe('scrollTop exceeds the height of the document', () => {
+    it('clamps the return value to 1', () => {
+      Object.defineProperties(document.documentElement, {
+        scrollTop: { value: 105, configurable: true },
+        scrollHeight: { value: 200, configurable: true },
+        clientHeight: { value: 100, configurable: true }
+      })
+
+      expect(scrollPercentY()).toBe(1)
+    })
+  })
+
+  describe('scrollTop is zero', () => {
+    it('returns 0', () => {
+      Object.defineProperties(document.documentElement, {
+        scrollTop: { value: 0, configurable: true },
+        scrollHeight: { value: 200, configurable: true },
+        clientHeight: { value: 100, configurable: true }
+      })
+
+      expect(scrollPercentY()).toBe(0)
+    })
+  })
+
+  describe('scrollTop is half way when scrollHeight equals clientHeight', () => {
+    it('returns 1', () => {
+      Object.defineProperties(document.documentElement, {
+        scrollTop: { value: 50, configurable: true },
+        scrollHeight: { value: 100, configurable: true },
+        clientHeight: { value: 100, configurable: true }
+      })
+
+      expect(scrollPercentY()).toBe(1.0)
+    })
```

Comment:
When I wrote this test it made me realise something wasn't quite right with how we're doing the "scroll amount" calculation.

I believe the more correct thing would be to do something like this:
```js
  const viewedArea = scrollTop + clientHeight
  const viewedPercent = viewedArea / scrollHeight
  const viewedPercentClamped = clamp(viewedPercent, 0, 1)
```

Doing it this way results in a lot less strange errors and appears to give more sensible numbers across the board.

The downside of changing it is that the current promos are configured to expect the current range of values, so without modifying the existing promos they would start behaving differently. 

Something to think about, not saying we necessarily need to rush into making immediate changes.